text/000-rustdoc-cfgs-handling.md

```diff
@@ -0,0 +1,305 @@
+Rustdoc: stabilization of the `doc(cfg*)` attributes
+
+- Features Name: `doc_cfg`
+- Start Date: 2022-12-07
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#43781](https://github.com/rust-lang/rust/issues/43781)
+
+
+# Summary
+[summary]: #summary
+
+This RFC aims at providing rustdoc users the possibility to add visual markers to the rendered documentation to know under which conditions an item is available (currently possible through the following unstable features: `doc_cfg`, `doc_auto_cfg` and `doc_cfg_hide`).
+
+It does not aim to allow having a same item with different `cfg`s to appear more than once in the generated documentation.
+
+It does not aim to document items which are *inactive* under the current configuration (i.e., “`cfg`'ed out”).
```

Comment:
...I'm slightly confused by this preamble, honestly, because I would figure the most important time to know when something is available is when it is, in fact, not available... when it has been "cfg'd out" but can be enabled by other means.

Reply:
You are right, this sentence doesn't tell the full story and should be improved (however, it's technically correct). I'm pretty sure the RFC does elaborate on it further down in the document (Ctrl+F `doc` (Match Case, Match Full Words)).

It implicitly refers to https://github.com/rust-lang/rust/issues/1998.

As you can probably guess, rustdoc doesn't and won't magically recompile the to-be-document crate under all possible configurations and combinations thereof to gather the information necessary to display the "portability info box".

So for the following crate, `function` won't show up in the generated docs unless you actually passed `--cfg special` to `rustdoc`:

```rs
#![feature(doc_auto_cfg)] // (A)
// #![feature(doc_cfg)] // (B)

// #[doc(cfg(special))] (B)
#[cfg(special)]
pub fn function() {}
```

Therefore, the common and offical workaround is the use of semi-special cfg (spec) `doc`:

```rs
#![feature(doc_auto_cfg)] // (A)
// #![feature(doc_cfg)] // (B)

// #[doc(cfg(special))] (B)
#[cfg(doc, special)]
pub fn function() {}
```

This gets more hairy if there are "multiple" functions called `function` (where the cfg specs differ for obvious reasons), one needs to choose with of the two functions to annotate with `doc`.

Of course, this doesn't scale to more complicated cases. E.g., if the parameter/return types differ by cfg spec. So, yeah. The general problem is not solved. That's what I meant with this sentence you highlighted.

Reply:
I mean, I've certainly written documentation that needs that invocation (or similar) to work properly. I guess I'm just feeling a bit perplexed that there is seemingly quite a lot of new concepts in this RFC and that's left unsolved. I would expect an RFC that has the sense of adding a Lot Of Stuff to also have an answer to that critical missing piece, because it's a right pain in the ass! ...or I would expect it to be smaller and more tightly scoped.

I realize that's a Pure Vibes thing, but I do think it can be a bit hard to evaluate a solution... and thus people would come to consensus slower... if it feels incorrectly scoped.

Reply:
> rustdoc doesn't and won't magically recompile the to-be-documented crate under all possible cfgs and combinations thereof

i mean, rustc now tracks enough information that it doesn’t have to
https://github.com/rust-lang/rust/pull/109005

Reply:
> i mean, rustc now tracks enough information that it doesn’t have to

Well, rustc only stores the *name* of cfg'ed out items but rustdoc would need access to so much more (struct fields, function signatures, ...). See [`StrippedCfgItem`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/expand/struct.StrippedCfgItem.html).

Reply:
There actually wasn't an explanation about this in the RFC so I added one in the `Unresolved questions` section.